src/archive/tar.rs

```diff
@@ -141,9 +145,7 @@ where
                 info(format!("Compressing '{}'", EscapedPathDisplay::new(path)));
             }
 
-            if path.is_dir() {
-                builder.append_dir(path, path)?;
-            } else if path.is_symlink() && !follow_symlinks {
+            if ((path.is_dir() && path.symlink_metadata()?.is_symlink()) || path.is_symlink()) && !follow_symlinks {
```

Comment:
The `path.is_dir() && path.symlink_metadata()?.is_symlink()` part seem to be redundant considering it's followed by `path.is_symlink()`.

You could cover both cases with 'is_symlink'

However, that function fails silently so, my suggestion is sticking to symlink_metadata to catch errors:

```suggestion
            if !follow_symlinks && path.symlink_metadata()?.is_symlink() {
```

I think that covers everything you wanted, and short-circuits before doing IO when follow_symlinks is true, makes sense?

Reply:
correct! I'd like your review style, which focuses on the performance.